wire-library/wire-runtime-swift/src/main/swift/ProtoEnum.swift

```diff
@@ -27,4 +29,120 @@ extension ProtoEnum {
         }
         self = result
     }
+
+    public init(from decoder: Decoder) throws {
+        // We support decoding from either the string value or the field index.
+        let container = try decoder.singleValueContainer()
+
+        if let string = try? container.decode(String.self) {
+            guard let value = Self(string) else {
+                throw ProtoDecoder.Error.unknownEnumString(type: Self.self, string: string)
+            }
+            self = value
+        } else {
+            // If the value wasn't a string, then look for the field index instead.
+            let fieldNumber = try container.decode(UInt32.self)
+            guard let value = Self(rawValue: fieldNumber) else {
+                throw ProtoDecoder.Error.unknownEnumCase(type: Self.self, fieldNumber: fieldNumber)
+            }
+
+            self = value
+        }
+    }
+
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.singleValueContainer()
+
+        switch encoder.enumEncodingStrategy {
+        case .string:
+            try container.encode(description)
+
+        case .integer:
+            try container.encode(rawValue)
+        }
+    }
+}
+
+/// This  type gives me access to `decoder.enumDecodingStrategy`
+private struct BoxedEnum<T: ProtoEnum> : Decodable {
+    let value: T?
+
+    init(value: T? = nil) {
+        self.value = value
+    }
+
+    init(from decoder: Decoder) throws {
+        let container = try decoder.singleValueContainer()
+
+        guard !container.decodeNil() else {
+            self.init()
+            return
+        }
+
+        switch decoder.enumDecodingStrategy {
+        case .shouldSkip:
+            let value = try? container.decode(T.self)
+            self.init(value: value)
+
+        case .shouldThrow:
+            let value = try container.decode(T.self)
+            self.init(value: value)
+        }
+    }
+}
+
+public extension KeyedDecodingContainer {
+    func decodeIfPresent<T: ProtoEnum>(
+        _: T.Type,
+        forKey key: Key
+    ) throws -> T? {
+        guard contains(key) else {
+            return nil
+        }
+
+        let box = try decode(BoxedEnum<T>.self, forKey: key)
+        return box.value
+    }
+
+    func decode<T: ProtoEnum>(
+        _: Array<T>.Type,
+        forKey key: Key
+    ) throws -> [T] {
+        var container = try nestedUnkeyedContainer(forKey: key)
+
+        var results: [T] = []
+        if let count = container.count {
+            results.reserveCapacity(count)
+        }
+
+        while !container.isAtEnd {
+            let box = try container.decode(BoxedEnum<T>.self)
+            if let value = box.value {
+                results.append(value)
+            }
+        }
+
+        return results
+    }
+
+    func decode<T: ProtoEnum>(
```

Comment:
It should be possible to do something like this for Dictionary; however, it's a bit more work and currently Wire doesn't handle them.

The only marginally hard part is getting access to `decoder.enumDecodingStrategy` to support lossy dictionary conversion.

Reply:
Broadly: 
* make another Box type for Dictionary<Enum, Decodable>
* make a hack coding key like `_DictionaryCodingKey`
* fetch a keyed container with the hack coding key
* `container.allKeys.reduce(into: [:])`
    * potentially lossily convert hack coding key -> enum
    * insert into memo

iOS 15.4 adds support for `CodingKeyRepresentable` but I don't know if that makes anything easier while still attempting to maintain lossy conversion.

Reply:
If you want me to implement this, let me know

Reply:
I ended up needing a property wrapper 🤷‍♂️